reclaim.c

```diff
@@ -876,3 +883,33 @@ GC_API void GC_CALL GC_enumerate_reachable_objects_inner(
   ed.client_data = client_data;
   GC_apply_to_all_blocks(GC_do_enumerate_reachable_objects, (word)&ed);
 }
+
+static void* init_finalize_thread(void *arg)
+{
+  while(1) {
+    pthread_mutex_lock(&flzr_mtx);
+    while (GC_should_invoke_finalizers() == 0) {
+      pthread_cond_wait(&flzr_t_has_work, &flzr_mtx);
+    }
+    pthread_mutex_unlock(&flzr_mtx);
```

Comment:
I *think* you might have a ping-pong situation here (warning: I might be wrong).

What I tend to do with condition variables (partly because I'm stupid and I find condvars hard) is associate a `bool` with the condvar. In essence, the condvar says "if true, there is work to do". Then I unset it in the "listening" thread so I end up with something like:

```c
  bool work_to_do = false; // this is protected by flzr_mtx
  while(1) {
    pthread_mutex_lock(&flzr_mtx);
    while (!work_to_do) {
      pthread_cond_wait(&flzr_t_has_work, &flzr_mtx);
    }
    work_todo = false;
    pthread_mutex_unlock(&flzr_mtx);
    // do actual work
}
```

At the moment, IIUC, you aren't doing the equivalent of `work_todo = false` with the mutex locked, so I *think* you might end up a situation where the finalisation thread keeps getting spuriously woken up even when there's no actual work to do.

Now, amongst the many caveats in my thinking is that I don't know how `GC_invoke_finalisers` synchronises with the GC thread(s). Maybe this is all handled elsewhere in a nice way.

Reply:
Er, I might be wrong about spurious wakeups: I think the problem is that you won't get woken up when you want.

Maybe this example in pizauth, which is very like what you've got here, might make sense of what I've said above. https://github.com/ltratt/pizauth/blob/master/src/server/notifier.rs#L33

Reply:
In this case `GC_should_invoke_finalizers()` is a function that returns 0 or 1 depending on if there's work to do. It is acting as the `work_todo` bool in this case. If there are no finalisers to be run, this will always return 0. 

`GC_should_invoke_finalizers` is essentially a wrapper around an atomic global so calling it is thread-safe. It only gets set to 1 from inside a collection (where the finalisers threads are suspended anyway). It gets unset by the finaliser thread when there are no more finalisers need to be run (again this is done atomically). I think this is the root of your issue -- should it be unset while we hold the lock?

Reply:
> GC_should_invoke_finalizers is essentially a wrapper around an atomic global

Aha!

> should it be unset while we hold the lock

Yes I think so.

Reply:
Try this: d281df7021b04938dfc09e93c502f875bb0e36fd

Reply:
I'm not convinced by the commit (or, at least, you'll need to convince me). I was expecting a global `bool` that the collector sets and the finaliser thread reads and then we'd have (roughly):

```c
static bool flzr_work_to_do = false;

fn f() {
  while(1) {
    pthread_mutex_lock(&flzr_mtx);
    while (!atomic_read(flzr_work_to_do)) {
      pthread_cond_wait(&cond_var, &flzr_mtx);
    }
    atomic_write(flzr_work_to_do, false);
    pthread_mutex_unlock(&flzr_mtx);
    // do actual work
  }
}
```

Instead, I think you have a race:

```c
    pthread_mutex_unlock(&flzr_mtx);
    GC_invoke_finalizers();
    // At this point the collector sets flzr_can_work to 1.
    pthread_mutex_lock(&flzr_mtx);
    // Now we clobber flzr_can_work!
    flzr_can_work = 0;
    pthread_mutex_unlock(&flzr_mtx);
```

Reply:
Right, I think I get it. `flzr_can_work` is a global bool (well `int`, but acting as a bool). The collector sets this while holding the lock: 

https://github.com/softdevteam/bdwgc/commit/d281df7021b04938dfc09e93c502f875bb0e36fd#diff-0fd3119c2bf4ac05eccc6aa476ae75d602b4e6acaf6a71a416acba6d8898fd6cR918 

But `flzr_can_work` needs to be unset before we actually do the work

Reply:
So this? 220bc1f83a9285881ee531abbf53af620403d1ad

Reply:
> But flzr_can_work needs to be unset before we actually do the work

Yep!

Reply:
@jacob-hughes Hmm, though this still feels really weird:

```c
flzr_can_work = GC_should_invoke_finalizers();
```

Doesn't that mean that `flzr_can_work` and whatever's behind `GC_should_invoke_finalizers` can get out of sync? I think we need to either (1) use the same var for both (2) set the vars at the same point for both `flzr_can_work` and `GC_should_invoke_finalizers`.

Reply:
I think this is now unnecessary. Yeah.

Reply:
I've just tested this: 56cc901f07a424294963a008f71b6bdbc0d2b192. Let me know if you're happy and I'll squash.

Reply:
Can you quickly check if the performance of `sum` is still the same or not with these changes?

Reply:
I can, but even with 10 iterations it will take about 30 mins to run :)

Reply:
Just check with 1 or 2? It'll probably be fairly obvious if this has any chance of changing performance.

Reply:
OK, offline, we think we see a difference.